src/gui/editors/SongEditor.cpp

```diff
@@ -531,41 +541,88 @@ void SongEditor::keyPressEvent( QKeyEvent * ke )
 }
 
 
-
-
 void SongEditor::wheelEvent( QWheelEvent * we )
 {
-	if( we->modifiers() & Qt::ControlModifier )
+	if ((we->modifiers() & Qt::ControlModifier) && (position(we).x() > m_trackHeadWidth))
 	{
-		int z = m_zoomingModel->value();
+		// calculate the min zoom based on song length and song editor size
+		int minZoom = calculateMinZoom();
+				
+		int z = m_zoomingLogModel->value();
+		if (we->angleDelta().y() > 0)
+			z = static_cast<int>(
+				std::ceil((std::floor(z / 100.0f * DEFAULT_PIXELS_PER_BAR) + 1) * 100.0f / DEFAULT_PIXELS_PER_BAR));
+		else
+			z = std::floor((std::ceil(z / 100.0f * DEFAULT_PIXELS_PER_BAR) - 1) * 100.0f / DEFAULT_PIXELS_PER_BAR);
+		//int z = scaleFunction(sli, true);
+		int pos = 0;
 
-		if(we->angleDelta().y() > 0)
-		{
-			z++;
-		}
-		else if(we->angleDelta().y() < 0)
+		if (z >= minZoom || we->angleDelta().y() > 0)
 		{
-			z--;
-		}
-		z = qBound( 0, z, m_zoomingModel->size() - 1 );
+			// if Control && Shift Modifiers, use predefinedZoom values
+			if (we->modifiers() & Qt::ShiftModifier)
+			{
+				int i = 0;
+				do
+				{
+					if (we->angleDelta().y() > 0)
+					{
+						// zooming out
+						pos = i;
+						if (PredefinedZoom[pos] > z) 
+						{ 
+							z = PredefinedZoom[std::min(pos, static_cast<int>(PredefinedZoom.size() - 1))];
+							break; 
+						}
+					}
+					else
+					{
+						// zooming in
+						pos = PredefinedZoom.size() - i - 1;
+						if (PredefinedZoom[pos] < z) 
+						{
+							z = PredefinedZoom[std::max(pos, 0)]; 
+							break;
+						}
+					}
+				} while (++i < PredefinedZoom.size());
+			}
 
+			z = std::clamp(z, minZoom, ZOOM_MAX);
 
-		int x = position(we).x() - m_trackHeadWidth;
-		// bar based on the mouse x-position where the scroll wheel was used
-		int bar = x / pixelsPerBar();
-		// what would be the bar in the new zoom level on the very same mouse x
-		int newBar = x / DEFAULT_PIXELS_PER_BAR / m_zoomLevels[z];
-		// scroll so the bar "selected" by the mouse x doesn't move on the screen
-		m_leftRightScroll->setValue(m_leftRightScroll->value() + bar - newBar);
+			int x = position(we).x() - m_trackHeadWidth;
+			// save the pixels up to the cursor next bar when starting zooming
+			if (!m_zvsStatus->isVisible()) { m_iniBar = static_cast<int>(x / pixelsPerBar()); }
```

Comment:
When zooming out and in quickly, you use `m_iniBar` to zoom back into the bar we were on before, instead of zooming in on the bar where the mouse is now.

This is a neat feature. But using TextFloat as a timer is confusing. And if you wait just a second, the feature breaks. IMO if the scroll speed is ok (so you don't accidentally zoom out too much) this is not a problem that needs to be fixed. Most other editors audio/video editors has this behavior.

Reply:
I removed TextFloat and reset m_iniBar in keyReleased as you mentions before.